symengine/eval_double.cpp

```diff
@@ -371,6 +371,27 @@ class EvalRealDoubleVisitor : public EvalDoubleVisitor<double, C>
         }
         result_ = result;
     };
+
+    void bvisit(const BooleanAtom &ba)
+    {
+        result_ = ba.get_val();
+    }
+
+    void bvisit(const Piecewise &pw)
+    {
+        if (neq(*pw.get_vec().back().second, *boolTrue)) {
+            throw SymEngineException(
+                "EvalDouble requires a (Expr, True) at the end of Piecewise");
+        }
```

Comment:
This check I think should be done as part of Piecewise Debug time check, and here we should simply assume it (perhaps with an assert statement).

This check should not run as part of the Release build.

Reply:
Here I simply mirrored how its done in [lambda_double.h](https://github.com/symengine/symengine/blob/master/symengine/lambda_double.h#L593-L596). Should I change both of them?

Reply:
We can change this in another PR after this is merged, but I think we should change it. @isuruf what do you think?

Reply:
If you'd be willing to merge https://github.com/symengine/symengine/pull/1501 then I can quickly add an assertion with a message.

Reply:
I've preemptively made these changes in 4eeeb5c, but of course they depend on #1501. Depending on your decision with that PR, I'll either rebase this branch when its merged or will change the check to a standard `SYMENGINE_ASSERT`.

Reply:
@certik, sympy allows not having a fallback condition and returns a `NaN` if that case is ever triggered.

Reply:
@isuruf I don't think I like that design in SymPy then. Is there a disadvantage of always requiring Piecewise to have (Expr, True) at the end?

@jppelteret I merged your PR. Feel free to ping us when tests pass and things are ready. It's time consuming for me to constantly keep checking things, but if you ping me, it's easy for me to do a final review & merge.

Reply:
> Is there a disadvantage of always requiring Piecewise to have (Expr, True) at the end?

Not compatible with sympy and more typing.

Reply:
@isuruf ok. What do you want to do here? If we want to allow this, then there should be a simple function in Piecewise returning true/false if this term is present, and then indeed this has to be checked in Release mode also.

Reply:
Might it not be sufficient just to return (or set `result=`) NaN instead of throwing at the end of these functions? That's the same as no condition being met, and would not require any check at the start of the evaluation function. Otherwise maybe one could inject the terminal criterion with result of NaN when the initial `Piecewise` is being created?

Reply:
> Otherwise maybe one could inject the terminal criterion with result of NaN when the initial Piecewise is being created?

Hmm, we could do this in symengine.py and always require `(Expr, True)` in C++ library.

Reply:
So I'm thinking that this could go [here](https://github.com/symengine/symengine/blob/master/symengine/logic.cpp#L126). What do you think?

Reply:
(i.e. augment `vec_` if there is no `boolTrue` criterion at the end of the input `vec`)

Reply:
Sounds good to me. We can do this in another PR though.

Reply:
I never liked returning NaN. In my opinion, there should be an exception right when this happens, or even not allowing it in the first place.

Reply:
What do you think about definitions like,  `Piecewise((-1, x < 0), (0, x == 0), (1, x > 0))` ?

Reply:
Are the two options `Piecewise((-1, x < 0), (0, x == 0), (1, x > 0))` vs `Piecewise((-1, x < 0), (0, x == 0), (1, True))`?

Reply:
Yes, should we allow the first or not?

Reply:
FWIW I would argue that `Piecewise((-1, x < 0), (0, x == 0), (1, x > 0))` is legitimate and should be allowed. To me, the second variant with the "catch all" seems to put a band-aid over potential logic issues with the conditions. I'd rather know that there's a gap in the intervals than somehow skip over them.

Reply:
Also, it does not preclude the user adding a terminal `(1,True)`, `(NaN, True)` (or whatever) if that's the desired logic. But the second variant is more restrictive.

Reply:
I suggest in this PR we don't change the current SymEngine behavior (whatever it is), and do a change in a separate PR.

I don't mind either one, but I would like to minimize extra checks needed in the code that uses Piecewise, for example the check above is one of such checks. So as far as I am concerned, I would like any such checks to be happening in the constructor in Debug mode.